redis-lettuce/src/main/java/io/micronaut/configuration/lettuce/cache/RedisConnectionPoolCache.java

```diff
@@ -0,0 +1,406 @@
+@Requires(classes = SyncCache.class, property = RedisSetting.REDIS_POOL)
+public class RedisConnectionPoolCache extends AbstractRedisCache<AsyncPool<StatefulConnection<byte[], byte[]>>> {
+    private final RedisAsyncCache asyncCache;
+    private final AsyncPool<StatefulConnection<byte[], byte[]>> asyncPool;
+
+    /**
+     * Creates a new redis cache for the given arguments.
+     *
+     * @param defaultRedisCacheConfiguration The default configuration
+     * @param redisCacheConfiguration        The configuration
+     * @param conversionService             The conversion service
+     * @param beanLocator                   The bean locator used to discover the redis connection from the configuration
+     * @param asyncPool                     Redis async pool
+     */
+    @SuppressWarnings("unchecked")
+    public RedisConnectionPoolCache(
+            DefaultRedisCacheConfiguration defaultRedisCacheConfiguration,
+            RedisCacheConfiguration redisCacheConfiguration,
+            ConversionService<?> conversionService,
+            BeanLocator beanLocator,
+            AsyncPool<StatefulConnection<byte[], byte[]>> asyncPool
+    ) {
+        super(defaultRedisCacheConfiguration, redisCacheConfiguration, conversionService, beanLocator);
+        this.asyncCache = new RedisAsyncCache();
+        this.asyncPool = asyncPool;
+    }
+
+    @Override
+    public String getName() {
+        return redisCacheConfiguration.getCacheName();
+    }
+
+    @Override
+    public AsyncPool<StatefulConnection<byte[], byte[]>> getNativeCache() {
+        return asyncPool;
+    }
+
+    @Override
+    public <T> T get(Object key, Argument<T> requiredType, Supplier<T> supplier) {
+        byte[] serializedKey = serializeKey(key);
+        try {
+            return asyncPool.acquire().thenCompose(connection -> {
+                try {
+                    RedisStringCommands<byte[], byte[]> commands = getRedisStringCommands(connection);
+                    return CompletableFuture.completedFuture(get(serializedKey, requiredType, supplier, commands));
+                } finally {
+                    asyncPool.release(connection);
+                }
+            }).get();
+        } catch (InterruptedException | ExecutionException e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    @Override
+    public void invalidate(Object key) {
+        byte[] serializedKey = serializeKey(key);
+        asyncPool.acquire().thenAccept(connection -> {
+            try {
+                RedisKeyCommands<byte[], byte[]> commands = getRedisKeyCommands(connection);
+                invalidate(Collections.singletonList(serializedKey), commands);
+            } finally {
+                asyncPool.release(connection);
+            }
+        }).join();
+    }
+
+    private void invalidate(List<byte[]> key, RedisKeyCommands<byte[], byte[]> commands) {
+        commands.del(key.toArray(new byte[0][]));
```

Comment:
Should this not use the pool?

Reply:
Hi!
I added a join around method which is calling the `invalidate` so now operation is synced and client will get control back only when item is invalidated. Can we resolve this issue?
Thanks.